File: navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/models/RouteLegProgress.java
package com.mapbox.services.android.navigation.v5.models;

import android.support.annotation.NonNull;

import com.mapbox.services.Constants;
import com.mapbox.services.Experimental;
import com.mapbox.services.api.ServicesException;
import com.mapbox.services.api.directions.v5.models.LegStep;
import com.mapbox.services.api.directions.v5.models.RouteLeg;
import com.mapbox.services.api.utils.turf.TurfConstants;
import com.mapbox.services.api.utils.turf.TurfMeasurement;
import com.mapbox.services.api.utils.turf.TurfMisc;
import com.mapbox.services.commons.geojson.LineString;
import com.mapbox.services.commons.geojson.Point;
import com.mapbox.services.commons.models.Position;
import com.mapbox.services.commons.utils.PolylineUtils;

import java.util.List;

@Experimental
public class RouteLegProgress {

  private RouteLeg routeLeg;
  private int stepIndex;
  private Position userSnappedPosition;
  private RouteStepProgress currentStepProgress;

  /**
   * Constructor for the route leg routeProgress information.
   *
   * @param routeLeg            The current {@link RouteLeg} the user is traversing along.
   * @param stepIndex           The current step index the user is on.
   * @param userSnappedPosition the users snapped location when routeProgress was last updated.
   * @since 0.1.0
   */
  public RouteLegProgress(@NonNull RouteLeg routeLeg, int stepIndex, @NonNull Position userSnappedPosition) {
    this.routeLeg = routeLeg;
    this.stepIndex = stepIndex;
    this.userSnappedPosition = userSnappedPosition;
    currentStepProgress = new RouteStepProgress(routeLeg, stepIndex, userSnappedPosition);
  }

  /**
   * Gives a {@link RouteStepProgress} object with information about the particular step the user is currently on.
   *
   * @return a {@link RouteStepProgress} object.
   * @since 0.1.0
   */
  public RouteStepProgress getCurrentStepProgress() {
    return currentStepProgress;
  }

  /**
   * Index representing the current step.
   *
   * @return an {@code integer} representing the current step the user is on.
   * @since 0.1.0
   */
  public int getStepIndex() {
    if (stepIndex < 0 || stepIndex > routeLeg.getSteps().size() - 1) {
      throw new ServicesException("RouteProgress step index is outside its index limit.");
    }
    return stepIndex;
  }

  /**
   * Total distance traveled in meters along current leg.
   *
   * @return a double value representing the total distance the user has traveled along the current leg, using unit
   * meters.
   * @since 0.1.0
   */
  public double getDistanceTraveled() {
    double distanceTraveled = routeLeg.getDistance() - getDistanceRemaining();
    if (distanceTraveled < 0) {
      distanceTraveled = 0;
    }
    return distanceTraveled;
  }

  /**
   * Provides the duration remaining in seconds till the user reaches the end of the route.
   *
   * @return {@code long} value representing the duration remaining till end of route, in unit seconds.
   * @since 0.1.0
   */
  public double getDistanceRemaining() {
    double distanceRemaining = 0;

    List<Position> coords = PolylineUtils.decode(getCurrentStep().getGeometry(), Constants.PRECISION_6);
    if (coords.size() > 1) {
      LineString slicedLine = TurfMisc.lineSlice(
        Point.fromCoordinates(userSnappedPosition),
        Point.fromCoordinates(coords.get(coords.size() - 1)),
        LineString.fromCoordinates(coords)
      );
      distanceRemaining += TurfMeasurement.lineDistance(slicedLine, TurfConstants.UNIT_METERS);
    }
    for (int i = stepIndex + 1; i < routeLeg.getSteps().size(); i++) {
      distanceRemaining += routeLeg.getSteps().get(i).getDistance();
    }

    return distanceRemaining;
  }

  /**
   * Provides the duration remaining in seconds till the user reaches the end of the current step.
   *
   * @return {@code long} value representing the duration remaining till end of step, in unit seconds.
   * @since 0.1.0
   */
  public double getDurationRemaining() {
    return (1 - getFractionTraveled()) * routeLeg.getDuration();
  }

  /**
   * Get the fraction traveled along the current leg, this is a float value between 0 and 1 and isn't guaranteed to
   * reach 1 before the user reaches the next leg (if another leg exist in route).
   *
   * @return a float value between 0 and 1 representing the fraction the user has traveled along the current leg.
   * @since 0.1.0
   */
  public float getFractionTraveled() {
    float fractionTraveled = 1;

    if (routeLeg.getDistance() > 0) {
      fractionTraveled = (float) (getDistanceTraveled() / routeLeg.getDistance());
      if (fractionTraveled < 0) {
        fractionTraveled = 0;
      }
    }
    return fractionTraveled;
  }

  /**
   * Get the previous step the user traversed along, if the user is still on the first step, this will return null.
   *
   * @return a {@link LegStep} representing the previous step the user was on, if still on first step in route, returns
   * null.
   * @since 0.1.0
   */
  public LegStep getPreviousStep() {
    if (getStepIndex() == 0) {
      return null;
    }
    return routeLeg.getSteps().get(getStepIndex() - 1);
  }

  /**
   * Returns the current step the user is traversing along.
   *
   * @return a {@link LegStep} representing the step the user is currently on.
   * @since 0.1.0
   */
  public LegStep getCurrentStep() {
    return routeLeg.getSteps().get(getStepIndex());
  }

  /**
   * Get the next/upcoming step immediately after the current step. If the user is on the last step on the last leg,
   * this will return null since a next step doesn't exist.
   *
   * @return a {@link LegStep} representing the next step the user will be on.
   * @since 0.1.0
   */
  public LegStep getUpComingStep() {
    if (routeLeg.getSteps().size() - 1 > getStepIndex()) {
      return routeLeg.getSteps().get(getStepIndex() + 1);
    }
    return null;
  }
}

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt
package com.mapbox.navigation.base.trip.model

import com.mapbox.api.directions.v5.models.LegStep
import com.mapbox.api.directions.v5.models.RouteLeg

/**
 * This is a progress object specific to the current leg the user is on. If there is only one leg
 * in the directions route, much of this information will be identical to the parent
 * [RouteProgress].
 *
 * The latest route leg progress object can be obtained through the [RouteProgressObserver].
 * Note that the route leg progress object's immutable.
 *
 * @param legIndex Index representing the current leg the user is on. If the directions route currently in use
 * contains more then two waypoints, the route is likely to have multiple legs representing the
 * distance between the two points.
 * @param routeLeg [RouteLeg] geometry
 * @param distanceTraveled Total distance traveled in meters along current leg
 * @param distanceRemaining The distance remaining in meters until the user reaches the end of the leg
 * @param durationRemaining The duration remaining in seconds until the user reaches the end of the current step
 * @param fractionTraveled The fraction traveled along the current leg, this is a float value between 0 and 1 and
 * isn't guaranteed to reach 1 before the user reaches the next waypoint
 * @param currentStepProgress [RouteStepProgress] object with information about the particular step the user
 * is currently on
 * @param upcomingStep Next/upcoming step immediately after the current step. If the user is on the last step
 * on the last leg, this will return null since a next step doesn't exist
 */
class RouteLegProgress internal constructor(
    val legIndex: Int,
    val routeLeg: RouteLeg?,
    val distanceTraveled: Float,
    val distanceRemaining: Float,
    val durationRemaining: Double,
    val fractionTraveled: Float,
    val currentStepProgress: RouteStepProgress?,
    val upcomingStep: LegStep?
) {

    /**
     * Regenerate whenever a change is made
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as RouteLegProgress

        if (legIndex != other.legIndex) return false
        if (routeLeg != other.routeLeg) return false
        if (distanceTraveled != other.distanceTraveled) return false
        if (distanceRemaining != other.distanceRemaining) return false
        if (durationRemaining != other.durationRemaining) return false
        if (fractionTraveled != other.fractionTraveled) return false
        if (currentStepProgress != other.currentStepProgress) return false
        if (upcomingStep != other.upcomingStep) return false

        return true
    }

    /**
     * Regenerate whenever a change is made
     */
    override fun hashCode(): Int {
        var result = legIndex
        result = 31 * result + (routeLeg?.hashCode() ?: 0)
        result = 31 * result + distanceTraveled.hashCode()
        result = 31 * result + distanceRemaining.hashCode()
        result = 31 * result + durationRemaining.hashCode()
        result = 31 * result + fractionTraveled.hashCode()
        result = 31 * result + (currentStepProgress?.hashCode() ?: 0)
        result = 31 * result + (upcomingStep?.hashCode() ?: 0)
        return result
    }

    /**
     * Returns a string representation of the object.
     */
    override fun toString(): String {
        return "RouteLegProgress(" +
            "legIndex=$legIndex, " +
            "routeLeg=$routeLeg, " +
            "distanceTraveled=$distanceTraveled, " +
            "distanceRemaining=$distanceRemaining, " +
            "durationRemaining=$durationRemaining, " +
            "fractionTraveled=$fractionTraveled, " +
            "currentStepProgress=$currentStepProgress, " +
            "upcomingStep=$upcomingStep" +
            ")"
    }
}
